sql/index.go

```diff
@@ -351,13 +351,17 @@ func (r *IndexRegistry) IndexesByTable(db, table string) []Index {
 // IndexByExpression returns an index by the given expression. It will return
 // nil it the index is not found. If more than one expression is given, all
 // of them must match for the index to be matched.
-func (r *IndexRegistry) IndexByExpression(db string, expr ...Expression) Index {
+func (r *IndexRegistry) IndexByExpression(db string, aliases map[string]string, expr ...Expression) Index {
 	r.mut.RLock()
 	defer r.mut.RUnlock()
 
-	expressions := make([]string, len(expr))
-	for i, e := range expr {
-		expressions[i] = e.String()
+	expressions := make(map[string]string)
```

Comment:
Why are we passing `expr` and `aliases` all the way here instead of just passing the expressions already normalized?

Instead of collecting a `map[string]string` of aliases, we could collect a `map[string]sql.Expression` and just pass the values of that map as `expr`. Then, no API would need to change.

Reply:
We need to pass aliases from `assignIndexes` to `getIndexes`, because how we can combine them into `map[string]sql.Expression` in an obvious way?
Moreover in `func getIndexes(e sql.Expression, a *Analyzer)` we cast expression and pass left/right to `getIndexes` again or to `IndexByExpression`.
What I can do is to combine expressions and aliases before calling `IndexByExpression` , so instead of having `IndexByExpression(db string, aliases map[string]string, expr ...Expression)`
we will have `IndexByExpression(db string, expressions map[string]string)`, so basically move outside the `IndexByExpression`:
```go
	expressions := make(map[string]string)
	for _, e := range expr {
		if aliases != nil {
			expressions[e.String()] = aliases[e.String()]
		} else {
			expressions[e.String()] = ""
		}
	}
```

Reply:
What I mean is not having to change the api of `IndexByExpression` or anything like that so they don't need to know anything about aliases.

Instead, just pass the dealiased expressions.

If you would have to pass `mytable.A` and `someAlias` to `IndexByExpression`, instead pass `mytable.A` and `foo(mytable.B)`, no need for any kind of mapping.

That way the functions that look for indexes don't need to know anything about aliases of any kind, just looking for those expressions.

Reply:
Ok, so I can normalize expressions and aliases and pass to IndexByExpression only expressions (or actually):
`IndexByExpression(db string, expr ...string) Index`
because we don't need whole `Expression` object inside `IndexByExpression` and we already have aliases as strings.

Reply:
But if you got the `string` of the expression, means you had the expression. Then instead I would pass the `sql.Expression` itself so the API does not need to change. The fact that we use the `string` is an implementation detail, it's better to keep passing `sql.Expression`s IMHO.

Reply:
@erizocosmico PTAL

Reply:
@erizocosmico - PTAL (rollback).